File: python/ch12_agent/agent_ch12/core/filesystem.py
"""工作区文件系统的领域错误和接口。

Java 对照：这里是领域异常 + `interface` 的组合。核心工具只依赖
`WorkspaceFileSystem`，不直接依赖 `pathlib` 或操作系统 API。
"""

from typing import Protocol


class WorkspacePathError(Exception):
    """路径不是工作区内的安全相对路径。

    这是什么：文件路径违反工作区边界的领域异常
    Java 类比：类似自定义的 PathEscapeException
    为什么需要：区分路径逃逸和其他文件错误，让权限层精确拒绝越界写入
    """


class TextNotFoundError(Exception):
    """编辑时找不到要求精确匹配的旧文本。

    这是什么：edit_file 操作中旧文本不存在的专用异常
    Java 类比：类似 TextNotFoundException
    为什么需要：避免盲目修改文件，edit 操作要求旧文本必须精确匹配
    """


class InvalidUtf8Error(Exception):
    """文件字节不能按严格 UTF-8 解码。

    这是什么：文件编码错误的领域异常
    Java 类比：类似 CharacterCodingException
    为什么需要：Agent 只处理 UTF-8 文本，避免把二进制文件或乱码发给模型
    """


class FileNotFoundError(Exception):
    """目标文件或目录不存在。

    这是什么：文件或目录缺失的领域异常
    Java 类比：类似 java.io.FileNotFoundException
    为什么需要：统一文件不存在的错误表示，避免直接暴露操作系统异常
    """


class InvalidFilePathError(Exception):
    """路径存在，但文件/目录类型不符合当前操作。

    这是什么：路径类型不匹配的领域异常
    Java 类比：类似 NotDirectoryException / NotRegularFileException
    为什么需要：区分"文件不存在"和"路径类型错误"（如把目录当文件读）
    """


class FileSystemOperationError(Exception):
    """无法归入其他领域错误的底层文件系统失败。

    这是什么：文件系统操作的兜底异常
    Java 类比：类似 IOException
    为什么需要：统一处理权限不足、磁盘满等无法精确分类的文件系统错误
    """


class WorkspaceWriteBoundary(Protocol):
    """权限层使用的窄接口，只判断写路径是否仍在工作区。

    这是什么：工作区边界检查的最小接口
    Java 类比：interface WorkspaceWriteBoundary { boolean isPathWithin(...); }
    为什么需要：权限策略只需要判断路径合法性，不需要完整文件系统能力
    """

    def is_path_within_workspace(self, workspace: str, relative_path: str) -> bool:
        """安全路径返回 True，路径逃逸返回 False，其他故障向上抛出。

        参数：
            workspace: 工作区根目录绝对路径
            relative_path: 待检查的相对路径

        返回：
            bool: 路径在工作区内返回 True，逃逸返回 False
        """


class WorkspaceFileSystem(WorkspaceWriteBoundary, Protocol):
    """工作区文件系统接口，类似 Java 的 `WorkspaceFileSystem` interface。

    这是什么：文件操作的领域边界接口
    Java 类比：interface WorkspaceFileSystem { String read(...); void write(...); }
    为什么需要：核心层不依赖 pathlib 或操作系统 API，测试时可用 Fake 替换
    """

    def read_file(self, workspace: str, relative_path: str, limit: int | None = None) -> str:
        """严格读取 UTF-8 文本；limit 按行数限制返回内容。

        参数：
            workspace: 工作区根目录
            relative_path: 工作区内的相对路径
            limit: 可选的行数限制，超出时追加剩余行数提示
        """

    def write_file(self, workspace: str, relative_path: str, content: str) -> int:
        """写入完整 UTF-8 文本并返回实际字节数。

        参数：
            workspace: 工作区根目录
            relative_path: 工作区内的相对路径
            content: 要写入的 UTF-8 文本内容
        """

    def edit_file(self, workspace: str, relative_path: str, old_text: str, new_text: str) -> None:
        """只替换第一次精确匹配，找不到时不修改文件。

        参数：
            workspace: 工作区根目录
            relative_path: 工作区内的相对路径
            old_text: 要替换的旧文本，必须精确匹配
            new_text: 替换后的新文本
        """

    def glob_files(self, workspace: str, pattern: str) -> tuple[str, ...]:
        """列出工作区内匹配模式的相对路径，并按稳定顺序返回。

        参数：
            workspace: 工作区根目录
            pattern: glob 模式，如 "**/*.py"
        """

File: typescript/chapters/ch02/src/core/filesystem.ts
/**
 * 文件系统领域错误与接口定义。
 * 每个错误类型对应一个精确的失败场景，工具层按 instanceof 映射为模型可见的稳定错误码。
 * WorkspaceFileSystem 只使用字符串路径和基本类型，不暴露 Node 实现细节。
 */
// 文件工具以细分错误类型区分路径逃逸、编码、缺失和 I/O 失败。
// 相对路径在词法或真实路径解析后逃出工作区时抛出。
export class WorkspacePathError extends Error {}

// edit_file 找不到要求的精确旧文本时抛出，保证文件没有被部分修改。
export class TextNotFoundError extends Error {}

// 文件字节不能按严格 UTF-8 解码时抛出，禁止静默替换损坏字节。
export class InvalidUtf8Error extends Error {}

// 目标文件或目录在执行操作前不存在时抛出。
export class FileNotFoundError extends Error {}

// 路径指向错误的文件类型，例如把目录作为文本文件时抛出。
export class InvalidFilePathError extends Error {}

// 无法归入其他领域错误的底层文件系统失败。
export class FileSystemOperationError extends Error {}

// 所有路径参数必须相对 workspace；实现负责在每次操作前保留此边界。
// 这是工具层唯一依赖的文件系统边界，Node 细节不进入 Agent Loop。
export interface WorkspaceFileSystem {
  // 读取严格 UTF-8 文本；limit 存在时按规范化行数截断输出。
  readFile(workspace: string, relativePath: string, limit?: number): Promise<string>;
  // 写入完整 UTF-8 内容并返回实际字节数，调用方据此向模型报告副作用。
  writeFile(workspace: string, relativePath: string, content: string): Promise<number>;
  // 仅替换第一次精确匹配，oldText 不存在时不写入任何变更。
  editFile(
    workspace: string,
    relativePath: string,
    oldText: string,
    newText: string,
  ): Promise<void>;
  // 按受限 glob 子集列举工作区相对路径，结果必须稳定排序。
  globFiles(workspace: string, pattern: string): Promise<readonly string[]>;
}
